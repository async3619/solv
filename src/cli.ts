import * as chalk from "chalk";
import { version } from "../package.json";

export function clearConsole() {
    process.stdout.write("\x1Bc");
}

export function drawLogo() {
    const ASCII_LOGO = `               __     
   _________  / /   __
  / ___/ __ \\/ / | / /
 (__  ) /_/ / /| |/ / 
/____/\\____/_/ |___/ ${chalk.italic.white(`v${version}`)}`;

    console.log(chalk.cyan(ASCII_LOGO));
}

export function breakLine(count = 1) {
    for (let i = 0; i < count; i++) {
        console.log();
    }
}

export function drawLine(length: number, char = "=") {
    console.log(new Array(length).fill(char).join(""));
}
