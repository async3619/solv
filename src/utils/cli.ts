import * as chalk from "chalk";
import { Command } from "commander";

import { serialize } from "./serialize";
import { CommandLineArgs, InputType } from "./types";

import { version, description } from "../../package.json";

export function clearConsole() {
    process.stdout.write("\x1Bc");
}

export function drawLogo() {
    const ASCII_LOGO = `               __     
   _________  / /   __
  / ___/ __ \\/ / | / /
 (__  ) /_/ / /| |/ / 
/____/\\____/_/ |___/ ${chalk.italic.white(`v${version}`)}`;

    console.log(chalk.cyan(ASCII_LOGO));
}

export function breakLine(count = 1) {
    for (let i = 0; i < count; i++) {
        console.log();
    }
}

export function drawLine(length: number, char = "=") {
    console.log(new Array(length).fill(char).join(""));
}

export async function parseCommandLine(argv: string[]): Promise<CommandLineArgs> {
    const program = await new Command()
        .name("solv")
        .description(description)
        .version(version)
        .argument("<url>", "specify a website url to solve")
        .option("--source, -s <source>", "specify source code path to watch")
        .option("--config, -c <config>", "specify configuration file path")
        .option("--no-overwrite, -n", "specify if program should not overwrite source code file")
        .option("--no-cache, -w", "specify if program should not cache challenge information")
        .parseAsync(argv);

    const { S: source, C: configPath, N: noOverwrite, W: noCache } = program.opts();
    const [targetUrl] = program.args;

    return {
        source,
        configPath,
        noCache,
        noOverwrite,
        targetUrl,
    };
}

export function renderSection(title: string, content: InputType, chalkFunction?: (target: string) => string) {
    let contentText: string | null;
    if (Array.isArray(content)) {
        contentText = content.map(item => serialize(JSON.parse(item))).join(" ");
    } else {
        contentText = content;
    }

    console.info(title);
    drawLine(15);

    const actualContent = contentText || chalk.italic("(empty)");
    console.info(chalkFunction ? chalkFunction(actualContent) : actualContent);

    breakLine();
}

export function getVersion() {
    return version;
}
