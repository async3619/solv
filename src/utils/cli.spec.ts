import { parseCommandLine, drawLine, drawLogo, breakLine, clearConsole } from "./cli";

describe("parseCommandLine", () => {
    it("should parse command line options properly", async () => {
        const data = await parseCommandLine([
            "",
            "solv",
            "-c",
            "./test/.solv.yml",
            "-s",
            "./test/test.ts",
            "-n",
            "https://example.com",
        ]);

        expect(data.source).toBe("./test/test.ts");
        expect(data.configPath).toBe("./test/.solv.yml");
        expect(data.noOverwrite).toBe(true);
        expect(data.targetUrl).toBe("https://example.com");
    });
});

describe("drawLine", () => {
    it("should draws a line with given length", () => {
        const buffer: string[] = [];
        console.log = (content: string) => buffer.push(content);

        drawLine(15);
        drawLine(5);

        expect(buffer[0]).toBe("===============");
        expect(buffer[1]).toBe("=====");
    });

    it("should draws a line with specific character", () => {
        const buffer: string[] = [];
        console.log = (content: string) => buffer.push(content);

        drawLine(15, "-");
        drawLine(5, "*");

        expect(buffer[0]).toBe("---------------");
        expect(buffer[1]).toBe("*****");
    });
});

describe("breakLine", function () {
    it("should breaks a line properly", function () {
        const buffer: string[] = [];
        console.log = (content: string) => buffer.push(content);

        breakLine();
        expect(buffer).toHaveLength(1);
    });

    it("should breaks lines depending on given count", function () {
        const buffer: string[] = [];
        console.log = (content: string) => buffer.push(content);

        breakLine(15);
        expect(buffer).toHaveLength(15);

        breakLine(5);
        expect(buffer).toHaveLength(20);
    });
});

describe("drawLogo", () => {
    it("should draws a logo properly", function () {
        const buffer: string[] = [];
        console.log = (content: string) => buffer.push(content);

        drawLogo();

        expect(buffer[0]).toMatchSnapshot();
    });
});

describe("clearConsole", function () {
    it("should clears console properly", function () {
        let data = "";
        (process.stdout.write as any) = (d: string) => (data = d);

        clearConsole();
        expect(data).toBe("\x1Bc");
    });
});
