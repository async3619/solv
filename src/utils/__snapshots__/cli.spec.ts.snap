// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`drawLogo should draws a logo properly 1`] = `
"[36m               __     [39m
[36m   _________  / /   __[39m
[36m  / ___/ __ \\/ / | / /[39m
[36m (__  ) /_/ / /| |/ / [39m
[36m/____/\\____/_/ |___/ [3m[37mv1.2.2[39m[36m[23m[39m"
`;
